examples/research_projects/zero-shot-distillation/distill_classifier.py

```diff
@@ -0,0 +1,320 @@
+
+    # Log on each process the small summary:
+    logger.warning(
+        f"Process rank: {training_args.local_rank}, device: {training_args.device}, n_gpu: {training_args.n_gpu}"
+        + f"distributed training: {bool(training_args.local_rank != -1)}, 16-bits training: {training_args.fp16}"
+    )
+    # Set the verbosity to info of the Transformers logger (on main process only):
+    if is_main_process(training_args.local_rank):
+        utils.logging.set_verbosity_info()
+        utils.logging.enable_default_handler()
+        utils.logging.enable_explicit_format()
+    logger.info(f"Training/evaluation parameters {training_args}")
+
+    # Set seed before initializing model.
+    set_seed(training_args.seed)
+
+    # 1. read in data
+    examples = read_lines(data_args.data_file)
+    class_names = read_lines(data_args.class_names_file)
+
+    # 2. get teacher predictions and load into dataset
+    logger.info("Generating predictions from zero-shot teacher model")
+    teacher_soft_preds = get_teacher_predictions(
+        teacher_args.teacher_name_or_path,
+        examples,
+        class_names,
+        teacher_args.hypothesis_template,
+        teacher_args.teacher_batch_size,
+        teacher_args.temperature,
+        teacher_args.multi_class,
+        data_args.use_fast_tokenizer,
+        training_args.no_cuda,
+    )
+    dataset = Dataset.from_dict(
+        {
+            "text": examples,
+            "labels": teacher_soft_preds,
+        }
+    )
+
+    # 3. create student
+    logger.info("Initializing student model")
+    model = AutoModelForSequenceClassification.from_pretrained(
+        student_args.student_name_or_path, num_labels=len(class_names)
+    )
+    tokenizer = AutoTokenizer.from_pretrained(student_args.student_name_or_path, use_fast=data_args.use_fast_tokenizer)
+    model.config.id2label = {i: label for i, label in enumerate(class_names)}
+    model.config.label2id = {label: i for i, label in enumerate(class_names)}
+
+    # 4. train student on teacher predictions
+    dataset = dataset.map(tokenizer, input_columns="text")
+    dataset.set_format("torch")
+
+    def compute_metrics(p, return_outputs=False):
+        preds = p.predictions.argmax(-1)
+        proxy_labels = p.label_ids.argmax(-1)  # "label_ids" are actually distributions
+        return {"agreement": (preds == proxy_labels).mean().item()}
+
+    trainer = DistillationTrainer(
+        model=model,
+        tokenizer=tokenizer,
+        args=training_args,
+        train_dataset=dataset,
+        compute_metrics=compute_metrics,
+    )
+
+    if training_args.do_train:
+        logger.info("Training student model on teacher predictions")
+        trainer.train()
+
+    if training_args.do_eval:
+        agreement = trainer.evaluate(eval_dataset=dataset)["eval_agreement"]
+        logger.info(f"Agreement of student and teacher predictions: {agreement * 100:0.2f}%")
+
+    trainer.save_model()
+
+
+if __name__ == "__main__":
+    main()
```

Comment:
TPU support could probably be added very easily by adding the following

```suggestion
def _mp_fn(index):
    # For xla_spawn (TPUs)
    main()
    
if __name__ == "__main__":
    main()
```

Reply:
This would work for the training step, but generating the pseudo-labels isn't done with Trainer so I'd have to add TPU support manually for that piece. I can do that if TPU inference is straightforward (I've yet to try it out), but otherwise I'd probably rather get this out there now and add support for other things like that in a separate PR.

Reply:
Looking at the rest of the script, it would work. Just each TPU would generate all the pseudo-labels. But that's also the case for multi-GPU at the moment.